### PR TITLE
Developer tooling improvements

### PR DESCRIPTION
**Pull request by str4d**
- `CircuitGates` tooling for printing circuit gate details
- `CircuitCost::marginal_proof_size` API
- Improvements to `CircuitLayout`

### CODE REVIEW

**Review by ebfull — APPROVED** [review]

File: src/dev/gates.rs [hunk 1]
```diff
@@ -0,0 +1,227 @@
...
+        let mut cs = ConstraintSystem::default();
+        let _ = C::configure(&mut cs);
+
+        let format_scalar = |s: F| {
+            // Format scalar as hex.
+            let s = format!("{:?}", s);
+            // Remove leading zeroes.
+            let s = s.strip_prefix("0x").unwrap();
+            let s = s.trim_start_matches('0');
+            format!("0x{}", s)
```
Comment (daira) [inline src/dev/gates.rs:107]: This formats 0 as just "0x".

Comment (str4d) [reply src/dev/gates.rs:107]: We never reach here with `0`; that constant is handled explicitly below.

Comment (str4d) [reply src/dev/gates.rs:107]: Hmm, actually we'd reach here if we multiplied an expression by the scalar 0, so we should fix this. But that seems more indicative of a circuit bug.

File: src/dev/gates.rs [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+                                } else {
+                                    format_scalar(s)
+                                }
+                            },
+                            &|selector| format!("S{}", selector.0),
+                            &|_, column, rotation| format!("F{}@{}", column, rotation.0),
+                            &|_, column, rotation| format!("A{}@{}", column, rotation.0),
+                            &|_, column, rotation| format!("I{}@{}", column, rotation.0),
+                            &|a| {
+                                if a.contains(' ') {
```
Comment (daira) [inline src/dev/gates.rs:138]: That's clever :-)

**Review by daira — COMMENTED** [review]
utACK

**Pull request merged**